discovery/gossiper.go

```diff
@@ -2411,91 +2300,243 @@ func (d *AuthenticatedGossiper) fetchNodeAnn(
+// goroutine that will resend any pending non-stale messages to the peer upon
+// subsequent reconnections.
+func (d *AuthenticatedGossiper) sendMsgReliably(msg lnwire.Message,
+	peerPubKey [33]byte) error {
+
+	// We'll add the message to the store to ensure we can resend it upon
+	// restarts if the peer has yet to come back online.
+	if err := d.cfg.MessageStore.AddMessage(msg, peerPubKey); err != nil {
+		return err
+	}
+
+	return d.sendMsg(msg, peerPubKey)
+}
+
+// sendMsg attempts to send a message to a peer.
+func (d *AuthenticatedGossiper) sendMsg(msg lnwire.Message,
+	compressedPeerPubKey [33]byte) error {
+
+	// If there doesn't exist a connection handler for this peer, we'll
+	// launch one now. This signals that we are not aware of the peer's
+	// connection status, so we'll request to be notified upon their
+	// connection and send the message then.
+	d.cachedPeersMtx.Lock()
+	_, ok := d.peerConnectionHandlers[compressedPeerPubKey]
+	d.cachedPeersMtx.Unlock()
+	if !ok {
+		peerPubKey, err := btcec.ParsePubKey(
+			compressedPeerPubKey[:], btcec.S256(),
+		)
 		if err != nil {
 			return err
 		}
 
-		// Encode the AnnounceSignatures message.
-		var b bytes.Buffer
-		if err := msg.Encode(&b, 0); err != nil {
-			return err
-		}
+		peerChan := make(chan lnpeer.Peer, 1)
+		d.cachedPeersMtx.Lock()
+		d.peerConnectionHandlers[compressedPeerPubKey] = peerChan
+		d.cachedPeersMtx.Unlock()
 
-		// Add the encoded message to the database using the peer
-		// + shortChanID as key.
-		return bucket.Put(key[:], b.Bytes())
+		d.wg.Add(1)
+		go d.peerConnectionHandler(
+			compressedPeerPubKey, peerPubKey, peerChan,
+		)
 
-	})
-	if err != nil {
-		return err
+		// Since this is the first time we register a listener for this
+		// peer, it'll automatically fire if they're connected, which
+		// will resend any pending messages. Since the message has
+		// already been written to disk, we can return early to avoid
+		// sending the message twice.
+		return nil
 	}
 
-	// We have succeeded adding the message to the database. We now launch
-	// a goroutine that will keep on trying sending the message to the
-	// remote peer until it succeeds, or the gossiper shuts down. In case
-	// of success, the message will be removed from the database.
-	d.wg.Add(1)
-	go func() {
-		defer d.wg.Done()
-		for {
-			log.Debugf("Sending AnnounceSignatures for channel "+
-				"%v to remote peer %x", msg.ChannelID,
-				remotePeer.SerializeCompressed())
-			err := d.cfg.SendToPeer(remotePeer, msg)
-			if err == nil {
-				// Sending succeeded, we can
-				// continue the flow.
-				break
+	// Otherwise, we'll check with our peer cache to determine if the peer
+	// is online. If they're not, a connection handler that requests for
+	// notifications upon reconnections already exists, which will send the
+	// message once fired, so there's nothing left for us to do.
+	d.cachedPeersMtx.Lock()
+	peer, ok := d.cachedPeers[compressedPeerPubKey]
```

Comment:
Hmm, looks like one of my comments regarding simplifying this got lost.. Basically I think caching peers like this are unnecessary, as the cache might get out of sync, leading to having to fall back waiting to them to become online anyway.

Instead, consider a simple loop like

```go

func peerHandler:
for { 
msg<-newMessages:
     store(msg)
<-comes online:
    send all stored messages
    break loop
}

for {
msg <- newMsgs
     store and send
<- goes offline:
     back to start
}

func sendMessage(msg):
      spin up peerHandler 
      newMsgs<-msg
```

Reply:
agree that doing this w/o a cache would be preferable to avoid getting out of sync

Reply:
Great suggestion @halseth, fixed!